editor/editor_tool_drawer.cpp

```diff
@@ -0,0 +1,496 @@
+
+	HBoxContainer *title_hb = memnew(HBoxContainer);
+	title_hb->set_alignment(BoxContainer::ALIGNMENT_CENTER);
+	title_panel->add_child(title_hb);
+
+	title_icon = memnew(TextureRect);
+	title_icon->set_stretch_mode(TextureRect::STRETCH_KEEP_CENTERED);
+	title_hb->add_child(title_icon);
+
+	title_label = memnew(Label);
+	title_label->set_vertical_alignment(VERTICAL_ALIGNMENT_CENTER);
+	title_hb->add_child(title_label);
+}
+
+void EditorToolDrawerGroup::set_title(const String p_title) {
+	title_label->set_text(p_title);
+}
+
+void EditorToolDrawerGroup::_notification(int p_notification) {
+	switch (p_notification) {
+		case NOTIFICATION_ENTER_TREE:
+		case NOTIFICATION_THEME_CHANGED: {
+			title_label->add_theme_color_override("font_color", get_theme_color(SNAME("group_title_color"), SNAME("EditorToolDrawer")));
+			title_label->add_theme_style_override("normal", get_theme_stylebox(SNAME("group_title"), SNAME("EditorToolDrawer")));
+		} break;
+	}
+}
+
+EditorToolDrawerGroup::EditorToolDrawerGroup() {
+	title_label = memnew(Label);
+	title_label->set_vertical_alignment(VERTICAL_ALIGNMENT_CENTER);
+	add_child(title_label);
+}
+
+void EditorToolDrawer::add_control_to_drawer(const String p_name, const Ref<Texture2D> &p_icon, Control *p_control) {
+	ERR_FAIL_COND(p_name.is_empty());
+	ERR_FAIL_COND(p_icon.is_null());
+	ERR_FAIL_COND(!p_control);
+
+	// Create the activator button for the drawer control.
+	Button *tb = memnew(Button);
+	tb->set_custom_minimum_size(Size2(24, 24) * EDSCALE);
+	tb->set_toggle_mode(true);
+	tb->set_tooltip(p_name);
+	tb->set_icon(p_icon);
+	tool_drawer_bar->add_child(tb);
+
+	Label *tb_label = memnew(Label);
+	tb_label->set_anchors_and_offsets_preset(LayoutPreset::PRESET_CENTER_LEFT);
+	tb_label->set_h_grow_direction(GROW_DIRECTION_BEGIN);
+	tb_label->set_v_grow_direction(GROW_DIRECTION_BOTH);
+	tb_label->set_vertical_alignment(VERTICAL_ALIGNMENT_CENTER);
+	tb_label->set_text(p_name);
+	tb->add_child(tb_label);
+
+	// Create the drawer section for the drawer control.
+	EditorToolDrawerSection *section = memnew(EditorToolDrawerSection);
+	section->set_title(p_name);
+	section->set_title_icon(p_icon);
+	section->hide();
+	tool_drawer->add_child(section);
+
+	section->add_content(p_control);
+
+	// Connect input signals.
+	tb->connect(SNAME("pressed"), callable_mp(this, &EditorToolDrawer::_toggle_section), varray(tb, section, p_control));
+
+	// Map the drawer control to the related nodes and data.
+	tool_drawer_button_map[p_control] = tb;
+	tool_drawer_section_map[p_control] = section;
+	tool_drawer_expanded_map[p_control] = false;
+
+	bool force_display_drawers = EDITOR_GET("editors/common/always_display_drawers");
+	if (force_display_drawers) {
+		tb->set_pressed(true);
+		section->set_visible(true);
+		tool_drawer_expanded_map[p_control] = true;
+	}
+
+	_update_mass_toggle_visibility();
+	_update_mass_toggle_icon();
+}
+
+void EditorToolDrawer::remove_control_from_drawer(Control *p_control) {
```

Comment:
Is this method needed? It's not used anywhere and the only use would be if we exposed drawers to scripting. Unless you already plan to do that.

Reply:
I do, yes.